pavelib/quality.py

```diff
@@ -301,48 +304,151 @@ def run_jshint(options):
 ])
 def run_safelint(options):
     """
     Runs safe_template_linter.py on the codebase
     """
 
-    violations_limit = int(getattr(options, 'limit', -1))
+    thresholds_option = getattr(options, 'thresholds', '{}')
+    try:
+        violation_thresholds = json.loads(thresholds_option)
+    except ValueError:
+        violation_thresholds = None
+    if isinstance(violation_thresholds, dict) is False or \
+            any(key not in ("total", "rules") for key in violation_thresholds.keys()):
+
+        raise BuildFailure(
+            """Error. Thresholds option "{thresholds_option}" was not supplied using proper format.\n"""
+            """Here is a properly formatted example, '{{"total":100,"rules":{{"javascript-escape":0}}}}' """
+            """with property names in double-quotes.""".format(
+                thresholds_option=thresholds_option
+            )
+        )
 
+    safelint_script = "safe_template_linter.py"
     safelint_report_dir = (Env.REPORT_DIR / "safelint")
     safelint_report = safelint_report_dir / "safelint.report"
     _prepare_report_dir(safelint_report_dir)
 
     sh(
-        "{repo_root}/scripts/safe_template_linter.py >> {safelint_report}".format(
+        "{repo_root}/scripts/{safelint_script} --rule-totals >> {safelint_report}".format(
             repo_root=Env.REPO_ROOT,
+            safelint_script=safelint_script,
             safelint_report=safelint_report,
         ),
         ignore_error=True
     )
 
+    safelint_counts = _get_safelint_counts(safelint_report)
+
     try:
-        num_violations = int(_get_count_from_last_line(safelint_report, "safelint"))
+        metrics_str = "Number of {safelint_script} violations: {num_violations}\n".format(
+            safelint_script=safelint_script, num_violations=int(safelint_counts['total'])
+        )
+        if 'rules' in safelint_counts and any(safelint_counts['rules']):
+            metrics_str += "\n"
+            for rule, count in safelint_counts['rules'].iteritems():
+                metrics_str += "{rule} violations: {count}\n".format(
+                    rule=rule,
+                    count=int(count)
+                )
     except TypeError:
         raise BuildFailure(
-            "Error. Number of safelint violations could not be found in {safelint_report}".format(
-                safelint_report=safelint_report
+            "Error. Number of {safelint_script} violations could not be found in {safelint_report}".format(
+                safelint_script=safelint_script, safelint_report=safelint_report
             )
         )
 
+    metrics_report = (Env.METRICS_DIR / "safelint")
     # Record the metric
-    _write_metric(num_violations, (Env.METRICS_DIR / "safelint"))
+    _write_metric(metrics_str, metrics_report)
+    # Print number of violations to log.
+    sh("cat {metrics_report}".format(metrics_report=metrics_report), ignore_error=True)
```

Comment:
@benpatterson When there was a violation this output went to the Jenkins console. I'm not sure how this works exactly, and if this is good enough.  I imagine the same would now be true with safecommit_report, but it doesn't yet have any violations, so that is untested.